File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/engine/preparation/specs/GivenACyclicQueue/WhenTheQueueIsStarted.java
package com.lasthopesoftware.bluewater.client.playback.engine.preparation.specs.GivenACyclicQueue;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.lasthopesoftware.bluewater.client.browsing.library.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.PreparedPlayableFileQueue;
import com.lasthopesoftware.bluewater.client.playback.file.preparation.PreparedPlayableFile;
import com.lasthopesoftware.bluewater.client.playback.file.preparation.queues.CyclicalFileQueueProvider;
import com.namehillsoftware.handoff.Messenger;
import com.namehillsoftware.handoff.promises.MessengerOperator;
import com.namehillsoftware.handoff.promises.Promise;
import com.namehillsoftware.handoff.promises.response.VoidResponse;

import junit.framework.Assert;

import org.junit.BeforeClass;
import org.junit.Test;

import java.util.List;
import java.util.Map;
import java.util.Random;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;

public class WhenTheQueueIsStarted {
	private static PreparedPlayableFileQueue queue;
	private static int startPosition;

	@BeforeClass
	public static void before() {
		final Random random = new Random(System.currentTimeMillis());
		final int numberOfFiles = random.nextInt(500);

		final List<ServiceFile> serviceFiles =
			Stream
				.range(0, numberOfFiles)
				.map(i -> new ServiceFile(random.nextInt()))
				.collect(Collectors.toList());

		Map<ServiceFile, MockResolveAction> fileActionMap =
			Stream
				.of(serviceFiles)
				.collect(Collectors.toMap(file -> file, file -> spy(new MockResolveAction())));

		final CyclicalFileQueueProvider bufferingPlaybackQueuesProvider
			= new CyclicalFileQueueProvider();

		startPosition = random.nextInt(numberOfFiles);

		queue =
			new PreparedPlayableFileQueue(
				() -> 1,
				(file, preparedAt) -> new Promise<>(fileActionMap.get(file)),
				bufferingPlaybackQueuesProvider.provideQueue(serviceFiles, startPosition));
	}

	@Test
	public void thenTheQueueStartsAtTheCorrectPosition() {
		queue
			.promiseNextPreparedPlaybackFile(0)
			.then(new VoidResponse<>(positionedPlaybackFile -> Assert.assertEquals(startPosition, positionedPlaybackFile.getPlaylistPosition())));
	}

	private static class MockResolveAction implements MessengerOperator<PreparedPlayableFile> {
		@Override
		public void send(Messenger<PreparedPlayableFile> resolve) {
			resolve.sendResolution(mock(PreparedPlayableFile.class));
		}
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/playback/engine/preparation/GivenACyclicQueue/WhenTheQueueIsStarted.kt
package com.lasthopesoftware.bluewater.client.playback.engine.preparation.GivenACyclicQueue

import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.PreparedPlayableFileQueue
import com.lasthopesoftware.bluewater.client.playback.file.PositionedPlayableFile
import com.lasthopesoftware.bluewater.client.playback.file.preparation.PreparedPlayableFile
import com.lasthopesoftware.bluewater.client.playback.file.preparation.queues.CyclicalFileQueueProvider
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.namehillsoftware.handoff.Messenger
import com.namehillsoftware.handoff.promises.MessengerOperator
import com.namehillsoftware.handoff.promises.Promise
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.joda.time.Duration
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import kotlin.random.Random.Default.nextInt

class WhenTheQueueIsStarted {

	private val mut by lazy {
		val numberOfFiles = nextInt(2, 500)
		val serviceFiles = (0..numberOfFiles).map { ServiceFile(nextInt()) }

		val fileActionMap = serviceFiles.associateBy ({ it }, { MockResolveAction() })
		val bufferingPlaybackQueuesProvider = CyclicalFileQueueProvider()
		val startPosition = nextInt(1, numberOfFiles)
		val queue = PreparedPlayableFileQueue(
			{ 1 },
			{ file, _ ->
				Promise(fileActionMap[file])
			},
			bufferingPlaybackQueuesProvider.provideQueue(serviceFiles, startPosition)
		)
		Pair(startPosition, queue)
	}

	private var positionedPlaybackFile: PositionedPlayableFile? = null

	@BeforeAll
	fun act() {
		val (_, queue) = mut
		positionedPlaybackFile = queue.promiseNextPreparedPlaybackFile(Duration.ZERO)?.toExpiringFuture()?.get()
	}

	@Test
	fun `then the queue starts at the correct position`() {
		assertThat(positionedPlaybackFile?.playlistPosition).isEqualTo(mut.first)
	}

	private class MockResolveAction : MessengerOperator<PreparedPlayableFile> {
		override fun send(resolve: Messenger<PreparedPlayableFile?>) {
			resolve.sendResolution(mockk(relaxed = true))
		}
	}
}
